src/routers/alpha-router/alpha-router.ts

```diff
@@ -1645,6 +1652,105 @@ export class AlphaRouter
       }
     }
 
+    let newSetCachedRoutesPath = false;
+    const shouldEnableCachedRoutesCacheInvalidationFix =
+      Math.random() * 100 <
+      (this.cachedRoutesCacheInvalidationFixRolloutPercentage ?? 0);
+
+    // we have to write cached routes right before checking swapRouteRaw is null or not
+    // because getCachedRoutes in routing-api do not use the blocks-to-live to filter out the expired routes at all
+    // there's a possibility the cachedRoutes is always populated, but swapRouteFromCache is always null, because we don't update cachedRoutes in this case at all,
+    // as long as it's within 24 hours sliding window TTL
+    if (shouldEnableCachedRoutesCacheInvalidationFix) {
+      // theoretically, when routingConfig.intent === INTENT.CACHING, optimisticCachedRoutes should be false
+      // so that we can always pass in cachedRoutes?.notExpired(await blockNumber, !routingConfig.optimisticCachedRoutes)
+      // but just to be safe, we just hardcode true when checking the cached routes expiry for write update
+      // we decide to not check cached routes expiry in the read path anyway
+      if (!cachedRoutes?.notExpired(await blockNumber, true)) {
+        // optimisticCachedRoutes === false means at routing-api level, we only want to set cached routes during intent=caching, not intent=quote
+        // this means during the online quote endpoint path, we should not reset cached routes
+        if (routingConfig.intent === INTENT.CACHING) {
```

Comment:
how often/ when is INTENT.CACHING sent?

Reply:
so every quote request (intent = quote) from upstream will have a corresponding routing lambda async caching https://github.com/Uniswap/routing-api/blob/08646b11ca1246ba4c666f4909661e234520a1ee/lib/handlers/router-entities/route-caching/dynamo-route-caching-provider.ts#L346.

Reply:
i see that based on [routing api/lib/handlers/router-entities/route-caching/dynamo-route-caching-provider.ts#L307](https://github.com/Uniswap/routing-api/blob/b1e24e225acf19b6fa3fa171d5a7d513dda25b55/lib/handlers/router-entities/route-caching/dynamo-route-caching-provider.ts#L307) this caching request is sent every few blocks (3 for mainnet and 15 for others) or if cache entry is missing. is this expected behavior?

Reply:
so yea, originally [filterExpiredCachedRoutes](https://github.com/Uniswap/routing-api/blob/b3025558e807c37e11ef87720b788bdc73d6f4ea/lib/handlers/router-entities/route-caching/dynamo-route-caching-provider.ts#L483) would filter against the expired cached routes, based on the block number that the cached routes got persisted into the dynamo DB. but with that bug in the expired cached routes, cached routes can remain as obsolete as thousands of blocks just on L1:

<img width="453" alt="Screenshot 2024-10-28 at 1 38 53 PM" src="https://github.com/user-attachments/assets/28be64f9-b615-4e62-a094-062c25523a30">

Reply:
now we have a change of plan as discussed last week, we will maintain the mistake in [filterExpiredCachedRoutes](https://github.com/Uniswap/routing-api/blob/b3025558e807c37e11ef87720b788bdc73d6f4ea/lib/handlers/router-entities/route-caching/dynamo-route-caching-provider.ts#L483), but instead do a cached routes update in the async path in SOR, so that we shouldn't anticipate a online quote endpoint latency increase.